### PR TITLE
Implement behavioural parity tests for Python vs Rust stream backends (BDD + unit)

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Implements behavioural parity tests to verify end-to-end parity between Python and Rust stream backends, covering empty streams, multi-byte UTF-8 at chunk boundaries, broken pipes, and backpressure scenarios. Includes unit tests, BDD-style tests, a shared parity test helper, and a Gherkin feature file. Adds an ExecPlan for behavioural parity tests and updates related docs.
- Adds support files for parity testing and updates to documentation and roadmap to reflect completion of the parity work.

## Changes
- New tests
  - cuprum/unittests/test_stream_parity.py: unit parity tests for fine-grained coverage
  - tests/behaviour/test_stream_parity_behaviour.py: BDD-style parity steps
  - tests/features/stream_parity.feature: Gherkin scenarios for edge-case parity
- New test helpers
  - tests/helpers/parity.py: shared parity catalogue, pipeline runner, and UTF-8 stress payload generator
- Documentation updates
  - docs/execplans/4-3-2-behavioural parity tests.md: ExecPlan for behavioural parity tests
  - docs/roadmap.md: mark 4.3.2 as complete
  - docs/users-guide.md: add parity guarantees for cross-backend consistency

## Rationale
- Ensures observable pipeline behavior is identical between Python and Rust backends without modifying production code.
- Uses a consistent test pattern (parity catalogue, allowlist, and end-to-end pipeline execution) to exercise the inter-stage dispatch path and the underlying backends.
- Provides deterministic UTF-8 payloads and large payload tests to stress chunking and backpressure handling.

## Test plan
- Run unit parity tests:
  - pytest cuprum/unittests/test_stream_parity.py
- Run behavioural parity tests (BDD):
  - pytest tests/behaviour/test_stream_parity_behaviour.py
- Run the feature file directly with pytest-bdd:
  - pytest tests/features/stream_parity.feature
- Ensure the Rust backend is skipped gracefully when the extension is unavailable.
- Full verification via make test as part of the project’s quality gates.

## How to run locally
- Export or set CUPRUM_STREAM_BACKEND to switch backends (Python or Rust).
- Run unit parity tests:
  - pytest cuprum/unittests/test_stream_parity.py
- Run BDD parity tests:
  - pytest tests/behaviour/test_stream_parity_behaviour.py
- Run the feature file via pytest-bdd:
  - pytest tests/features/stream_parity.feature

## Compatibility and constraints
- No production source code changes; tests exercise existing public interfaces and the runtime dispatch path.
- Parity tests rely on the project’s existing test harness (scoped execution, allowlists, and pipeline.run_sync).
- If the Rust extension is unavailable, related tests are skipped to avoid failures.

## Artifacts
Files added:
- cuprum/unittests/test_stream_parity.py
- tests/behaviour/test_stream_parity_behaviour.py
- tests/features/stream_parity.feature
- tests/helpers/parity.py
Files updated:
- docs/execplans/4-3-2-behavioural parity tests.md
- docs/roadmap.md
- docs/users-guide.md

## Context and references
- Task: https://www.devboxer.com/task/07d6e6fa-c6ae-4ca8-95ea-de6f6ab04b26

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[cuprum/unittests/test_stream_parity.py](https://github.com/leynos/cuprum/pull/40/files/c60f2ad877f3b911fded15e95f2ce9b013730ff5#diff-a2945514c43502bd21f88c296f38395b3416f4738d9e780868e19aec0e16e54d)

Comment on lines +226 to +246

~~~python
def test_backpressure_large_payload_two_stages(
    stream_backend: str,
) -> None:
    """Two-stage pipeline preserves a 1 MB payload.

    Parameters
    ----------
    stream_backend : str
        The active stream backend (injected by fixture).
    """
    size = 1024 * 1024
    # Generate payload inside the subprocess to stay within the OS
    # command-line length limit.
    script = f"import sys; sys.stdout.write('x' * {size})"
    pipeline, allowlist = _build_pipeline(script)
    result = run_parity_pipeline(pipeline, allowlist)

    assert result.stdout == "x" * size
    assert result.ok is True
    assert len(result.stages) == 2
    assert all(s.exit_code == 0 for s in result.stages)
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=40&biomarker=Code+Duplication&filename=cuprum%2Funittests%2Ftest_stream_parity.py&method=)
The module contains 6 functions with similar structure: test_backpressure_large_payload_three_stages,test_backpressure_large_payload_two_stages,test_mixed_utf8_payload_survives_pipeline,test_utf8_four_byte_chars_survive_pipeline and 2 more functions

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[tests/behaviour/test_stream_parity_behaviour.py](https://github.com/leynos/cuprum/pull/40/files/c60f2ad877f3b911fded15e95f2ce9b013730ff5#diff-16a3df10a63f369e12be7c008daa888c5923f670827d63d3770953c47cd41000)

Comment on lines +266 to +276

~~~python
def then_stdout_large(pipeline_result: PipelineResult) -> None:
    """Assert that pipeline stdout matches the 1 MB payload.

    Parameters
    ----------
    pipeline_result : PipelineResult
        The result from pipeline execution.
    """
    assert pipeline_result.stdout == _LARGE_PAYLOAD, (
        "expected 1 MB payload to survive pipeline intact"
    )
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=40&biomarker=Code+Duplication&filename=tests%2Fbehaviour%2Ftest_stream_parity_behaviour.py&method=)
The module contains 3 functions with similar structure: then_stdout_empty,then_stdout_large,then_stdout_utf8

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[tests/behaviour/test_stream_parity_behaviour.py](https://github.com/leynos/cuprum/pull/40/files/f6dd6af7cf136da6e207dcf273c647cccc29d6d1#diff-16a3df10a63f369e12be7c008daa888c5923f670827d63d3770953c47cd41000)

Comment on lines +356 to +369

~~~python
def then_stdout_large(
    pipeline_result: PipelineResult,
    large_payload: str,
) -> None:
    """Assert that pipeline stdout matches the 1 MB payload.

    Parameters
    ----------
    pipeline_result : PipelineResult
        The result from pipeline execution.
    large_payload : str
        Session-scoped 1 MB payload fixture.
    """
    _assert_stdout_matches(pipeline_result, large_payload, "1 MB payload")
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=40&biomarker=Code+Duplication&filename=tests%2Fbehaviour%2Ftest_stream_parity_behaviour.py&method=)
The module contains 2 functions with similar structure: then_stdout_large,then_stdout_utf8

File: tests/helpers/parity.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    Returns
+    -------
+    PipelineResult
+        The result of synchronous pipeline execution.
+    """
+    with scoped(ScopeConfig(allowlist=allowlist)):
+        return pipeline.run_sync()
+
+
+def utf8_stress_payload(n_chars: int = 8192) -> str:
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/helpers/parity.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: cuprum/unittests/test_stream_parity.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    )
+    downstream_script = "import sys; sys.stdout.write(sys.stdin.read(10))"
+    pipeline = python_cmd("-c", upstream_script) | python_cmd("-c", downstream_script)
+    allowlist = frozenset([python_prog])
+    result = run_parity_pipeline(pipeline, allowlist)
+
+    # The pipeline must complete (no deadlock). If it hangs,
+    # pytest-timeout kills the test. The downstream reads 10 bytes.
+    assert result.stages is not None
+    assert len(result.stages) == 2
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_stream_parity.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/helpers/parity.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    n_chars : int
+        Number of characters to generate. Defaults to ``8192``.
+
+    Returns
+    -------
+    str
+        A string containing a mix of 1, 2, 3, and 4-byte UTF-8
+        characters.
+    """
+    rng = random.Random(_SEED)  # noqa: S311
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/helpers/parity.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai review

File: cuprum/unittests/test_stream_parity.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_stream_parity.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Code Duplication
tests/behaviour/test_stream_parity_behaviour.py: 

### What lead to degradation?
The module contains 2 functions with similar structure: then_stdout_large,then_stdout_utf8

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.
